GCEWindowsAgent/addresses.go

```diff
@@ -17,32 +17,77 @@ package main
 import (
 	"net"
 	"reflect"
+	"strconv"
 	"strings"
 
 	"github.com/GoogleCloudPlatform/compute-image-windows/logger"
+	"github.com/go-ini/ini"
 )
 
-var addressDisabled = false
-var addressKey = regKeyBase + `\ForwardedIps`
+var (
+	addressDisabled  = false
+	addressKey       = regKeyBase + `\ForwardedIps`
+	oldWSFCAddresses string
+	oldWSFCEnable    bool
+)
 
 type addresses struct {
 	newMetadata, oldMetadata *metadataJSON
+	config                   *ini.File
 }
 
 func (a *addresses) diff() bool {
-	return !reflect.DeepEqual(a.newMetadata.Instance.NetworkInterfaces, a.oldMetadata.Instance.NetworkInterfaces) ||
-		a.newMetadata.Instance.Attributes.WSFCAddresses != a.oldMetadata.Instance.Attributes.WSFCAddresses ||
-		a.newMetadata.Instance.Attributes.EnableWSFC != a.oldMetadata.Instance.Attributes.EnableWSFC
-}
+	wsfcAddresses := a.config.Section("wsfc").Key("addresses").String()
+	if len(wsfcAddresses) == 0 {
+		if len(a.newMetadata.Instance.Attributes.WSFCAddresses) > 0 {
+			wsfcAddresses = a.newMetadata.Instance.Attributes.WSFCAddresses
+		} else if len(a.newMetadata.Project.Attributes.WSFCAddresses) > 0 {
+			wsfcAddresses = a.newMetadata.Project.Attributes.WSFCAddresses
+		}
+	}
 
-func (a *addresses) disabled() bool {
-	d := a.newMetadata.Instance.Attributes.DisableAddressManager
-	if d != addressDisabled {
-		addressDisabled = d
-		logStatus("address", d)
+	wsfcEnable, err := a.config.Section("wsfc").Key("enable").Bool()
+	if err != nil {
+		wsfcEnable, err = strconv.ParseBool(a.newMetadata.Instance.Attributes.EnableWSFC)
+		if err != nil {
+			wsfcEnable, err = strconv.ParseBool(a.newMetadata.Project.Attributes.EnableWSFC)
+			if err != nil {
+				wsfcEnable = false
+			}
+		}
 	}
 
-	return d
+	diff := !reflect.DeepEqual(a.newMetadata.Instance.NetworkInterfaces, a.oldMetadata.Instance.NetworkInterfaces) ||
+		wsfcEnable != oldWSFCEnable || wsfcAddresses != oldWSFCAddresses
+
+	oldWSFCAddresses = wsfcAddresses
+	oldWSFCEnable = wsfcEnable
+	return diff
+}
+
+func (a *addresses) disabled() (disabled bool) {
```

Comment:
Would it be worth adding this function to a common lib instead of repeating it? I don't know if there is an easy way to specify the metadata key as an argument. If you can think of one, that approach might make sense.